python/examples/bare_matmul.py

```diff
@@ -0,0 +1,39 @@
+import torch
+
+import triton
+import triton.language as tl
+import benchmark
+
+
+@triton.jit
+def bare_matmul(X, Y, Z, BLOCK_SIZE: tl.constexpr):
+    pid = tl.program_id(0)
+    
+    offs_x = pid * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
+    offs_y = tl.arange(0, BLOCK_SIZE)
+
+    x = tl.load(X + offs_x[:, None])
+    y = tl.load(Y + offs_y[None, :])
+
+    z = tl.dot(x, y)
+    tl.store(Z + offs_x[:, None] + offs_y[None, :], z)
+
```

Comment:
I believe there's a mistake in this triton program. We're always doing matmul of size $[BLOCK_SIZE, 1] @ [1, BLOCK_SIZE]$ which is $[N, 1] @ [1, N]$ in the triton version. But the torch version is doing $[N, N] @ [N, N]$. Perhaps this is why bare_matmul doesn't show any speed up at all?

Reply:
oh, thanks a lot for noticing, let me retry and get back to you!

Reply:
thanks again for noticing, I have fixed the kernel and the difference is very notable:

```
> python bare_matmul.py
bench_matmul(128, 'test') {}, 20 times, all results in seconds
Avg=0.035736, min=0.003327, 20pp=0.004988, 50pp=0.005001, 90pp=0.005019, max=0.621418
bench_matmul(128, 'torch') {}, 20 times, all results in seconds
Avg=0.000406, min=0.000218, 20pp=0.000219, 50pp=0.000219, 90pp=0.000403, max=0.002592
bench_matmul(128, 'triton') {}, 20 times, all results in seconds
Avg=0.001747, min=0.001088, 20pp=0.001148, 50pp=0.001160, 90pp=0.003361, max=0.003988
bench_matmul(256, 'test') {}, 20 times, all results in seconds
Avg=0.049110, min=0.005234, 20pp=0.005498, 50pp=0.006050, 90pp=0.019442, max=0.808195
bench_matmul(256, 'torch') {}, 20 times, all results in seconds
Avg=0.002304, min=0.000897, 20pp=0.000945, 50pp=0.002550, 90pp=0.003497, max=0.005217
bench_matmul(256, 'triton') {}, 20 times, all results in seconds
Avg=0.003742, min=0.003167, 20pp=0.003176, 50pp=0.003181, 90pp=0.005196, max=0.006016
bench_matmul(512, 'test') {}, 20 times, all results in seconds
Avg=0.076103, min=0.023856, 20pp=0.029637, 50pp=0.030622, 90pp=0.062977, max=0.842385
bench_matmul(512, 'torch') {}, 20 times, all results in seconds
Avg=0.005920, min=0.004630, 20pp=0.004641, 50pp=0.005528, 90pp=0.007751, max=0.011366
bench_matmul(512, 'triton') {}, 20 times, all results in seconds
Avg=0.011808, min=0.011642, 20pp=0.011654, 50pp=0.011680, 90pp=0.011719, max=0.014201
> export TRITON_SHARED_USE_OPENBLAS=
> python bare_matmul.py
bench_matmul(128, 'test') {}, 20 times, all results in seconds
Avg=0.075897, min=0.006302, 20pp=0.006325, 50pp=0.006347, 90pp=0.006507, max=1.396113
bench_matmul(128, 'torch') {}, 20 times, all results in seconds
Avg=0.000225, min=0.000218, 20pp=0.000218, 50pp=0.000219, 90pp=0.000230, max=0.000269
bench_matmul(128, 'triton') {}, 20 times, all results in seconds
Avg=0.005761, min=0.005674, 20pp=0.005698, 50pp=0.005723, 90pp=0.005954, max=0.005984
bench_matmul(256, 'test') {}, 20 times, all results in seconds
Avg=0.083720, min=0.045208, 20pp=0.045586, 50pp=0.045785, 90pp=0.046000, max=0.805519
bench_matmul(256, 'torch') {}, 20 times, all results in seconds
Avg=0.000898, min=0.000891, 20pp=0.000896, 50pp=0.000896, 90pp=0.000901, max=0.000925
bench_matmul(256, 'triton') {}, 20 times, all results in seconds
Avg=0.044448, min=0.043970, 20pp=0.044233, 50pp=0.044460, 90pp=0.044648, max=0.045219
bench_matmul(512, 'test') {}, 20 times, all results in seconds
Avg=0.395702, min=0.351327, 20pp=0.356543, 50pp=0.360086, 90pp=0.362602, max=1.096930
bench_matmul(512, 'torch') {}, 20 times, all results in seconds
Avg=0.004640, min=0.004615, 20pp=0.004621, 50pp=0.004629, 90pp=0.004654, max=0.004780
bench_matmul(512, 'triton') {}, 20 times, all results in seconds
Avg=0.348616, min=0.346798, 20pp=0.347046, 50pp=0.347344, 90pp=0.353371, max=0.354604
```